### PR TITLE
Faster sparse_dense on GPUs

### PR DESCRIPTION
**Pull request by tkonolige**
I've written a faster sparse_dense for GPUs using tir. This sparse_dense requires a padded matrix, so I've added a new op sparse_dense_padded. AlterOpLayout should transform sparse_dense to sparse_dense_padded when using a gpu.

This new sparse_dense improves prunebert performance from 155.41ms mean to 7.75ms mean. In general, this implementation is faster than cublas dense on matrices with density < 0.05 and is often faster than cusparse for machine learning workloads.

### CODE REVIEW

**Comment by ANSHUMAN87** [comment]
> I've written a faster sparse_dense for GPUs using tir. This sparse_dense requires a padded matrix, so I've added a new op sparse_dense_padded. AlterOpLayout should transform sparse_dense to sparse_dense_padded when using a gpu.
> 
> This new sparse_dense improves prunebert performance from 155.41ms mean to 7.75ms mean. In general, this implementation is faster than cublas dense on matrices with density < 0.05 and is often faster than cusparse for machine learning workloads.

@tkonolige : Thanks for the PR! The data looks quite impressive :+1: 
I was wondering whether we can add some sort of benchmark testcase here , tuned to your shared data?

**Comment by tkonolige** [comment]
@ANSHUMAN87 Right now TVM does not do any testing for performance regressions. The hard part in setting up performance testing is that is varies from run to run and machine to machine.

**Comment by ANSHUMAN87** [comment]
@tkonolige : I understand your concern clearly. However it was just a thought. Even if run to run or machine to machine difference, the relative reference would be same. But may be we don't have to do as part of this PR :)
I will go through deep into your PR, will share my comment if any. Thanks!

**Comment by mkstarr** [comment]
Have you considered your syntax errors?

**Comment by tkonolige** [comment]
@vinx13 @antinucleon @Laurawly @jwfromm @ajtulloch I think this is ready for review.

File: src/relay/op/nn/sparse.cc [hunk 1]
```diff
@@ -82,6 +82,35 @@ RELAY_REGISTER_OP("nn.sparse_dense")
...
+}
+
+TVM_REGISTER_GLOBAL("relay.op.nn._make.sparse_dense_padded")
+    .set_body([](const TVMArgs& args, TVMRetValue* rv) {
+      runtime::detail::unpack_call<Expr, 4>(MakeSparseDensePadded, args, rv);
+    });
+
+RELAY_REGISTER_OP("nn.sparse_dense_padded")
+    .describe(
+        R"code(Applies a sparse linear transformation: :math:`Y = XW^T` with X sparse. This variation uses a matrix with row lengths padded to a multiple of 32 for better GPU performance.
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply src/relay/op/nn/sparse.cc]: Good catch, fixed it for sparse_dense too.

File: python/tvm/topi/cuda/sparse.py [hunk 2]
```diff
@@ -97,3 +94,287 @@ def _callback(op):
...
+    Roughly follows this:
+    https://developer.nvidia.com/blog/efficient-matrix-transpose-cuda-cc/, but
+    without the padding for shared memory. For better performance, we could
+    rewrite it in tir to add the padding.
+    """
+
+    def _callback(op):
+        # pylint: disable=invalid-name
+        m, n = s[op].op.axis
+        warp_size = int(tvm.target.Target.current(allow_none=False).thread_warp_size)
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply python/tvm/topi/cuda/sparse.py:110]: Yes, 4 warps per block is slightly faster than 1 warp per block. I've added a comment to this effect.

File: python/tvm/tir/ir_builder.py [hunk 3]
```diff
@@ -74,8 +75,23 @@ def asobject(self):
...
+        dim_size = 1
+        lidx = 0
+        for dim, idx in zip(reversed(self._shape), reversed(index)):
+            lidx += idx * dim_size
+            dim_size *= dim
+        return lidx
+
     def __getitem__(self, index):
         t = DataType(self._content_type)
+        index = self._linear_index(index)
```
Comment (ANSHUMAN87) [inline python/tvm/tir/ir_builder.py:94]: @tkonolige : I am just trying to understand why this change is required for your requirement? 
Would you please help me understand. Thanks!

File: python/tvm/tir/ir_builder.py [hunk 4]
```diff
@@ -62,9 +62,10 @@ class BufferVar(ObjectGeneric):
     IRBuilder.allocate
     """
 
-    def __init__(self, builder, buffer_var, content_type):
+    def __init__(self, builder, buffer_var, shape, content_type):
```
Comment (ANSHUMAN87) [inline python/tvm/tir/ir_builder.py:65]: I don't think this change is necessary, but i would wait for your response for my other comment. Maybe after that we can discuss here :)

File: python/tvm/tir/ir_builder.py (hunk 3 above)
Comment (tkonolige) [reply python/tvm/tir/ir_builder.py:94]: This allows you to use multidimensional indices with IRBuilder. Manually calculating the correct index into multidimensional arrays caused a couple of bugs when I was developing this code. Using multidimensional indices made everything cleaner and easier to debug.

Comment (ANSHUMAN87) [reply python/tvm/tir/ir_builder.py:94]: Thanks @tkonolige for clarification!

File: python/tvm/tir/ir_builder.py (hunk 4 above)
Comment (ANSHUMAN87) [reply python/tvm/tir/ir_builder.py:65]: I have one query here. What happens if we don't have the shape info. In that case, how do you enforce the check in _linear_index ?

For example: We create a buffer pointer out of certain existing buffer, in that case you may not have necessarily the shape info, in that case, how it works ?

Comment (ANSHUMAN87) [reply python/tvm/tir/ir_builder.py:65]: This is change okay. Post clarification!

Comment (tkonolige) [reply python/tvm/tir/ir_builder.py:65]: You can use shape=None and the code will work fine assuming you always use a single index.

File: python/tvm/tir/ir_builder.py [hunk 5]
```diff
@@ -74,8 +75,23 @@ def asobject(self):
     def dtype(self):
         return self._content_type
 
+    def _linear_index(self, index):
+        if not isinstance(index, tuple):
```
Comment (ANSHUMAN87) [inline python/tvm/tir/ir_builder.py:79]: We can also check for _shape is None, then return here.

File: python/tvm/tir/ir_builder.py (hunk 4 above)
Comment (ANSHUMAN87) [reply python/tvm/tir/ir_builder.py:65]: Request to add small comment section to describe new way to use it.
As the new cases are 2, either with already linear index or actual index. It will help understand the new change better.

File: python/tvm/topi/nn/sparse.py [hunk 6]
```diff
@@ -207,3 +207,28 @@ def _csr_transpose_ir(data, indices, indptr, out_data, out_indices, out_indptr):
         last[0] = temp2[0]
 
     return irb.get()
+
+
+@tvm.target.generic_func
+def sparse_dense_alter_layout(_attrs, _inputs, _tinfos, _out_type):
```
Comment (ANSHUMAN87) [inline python/tvm/topi/nn/sparse.py:213]: Any reason to use _ naming for inputs?

File: src/relay/op/nn/sparse.cc [hunk 7]
```diff
@@ -76,7 +76,36 @@ TVM_REGISTER_GLOBAL("relay.op.nn._make.sparse_dense")
...
+  static const Op& op = Op::Get("nn.sparse_dense_padded");
+  return Call(op, {data, weight_data, weight_indices, weight_indptr}, Attrs(attrs), {});
+}
+
+TVM_REGISTER_GLOBAL("relay.op.nn._make.sparse_dense_padded")
+    .set_body([](const TVMArgs& args, TVMRetValue* rv) {
+      runtime::detail::unpack_call<Expr, 4>(MakeSparseDensePadded, args, rv);
+    });
+
+RELAY_REGISTER_OP("nn.sparse_dense_padded")
```
Comment (ANSHUMAN87) [inline src/relay/op/nn/sparse.cc:106]: In which case this Op will be used independently?

File: python/tvm/topi/nn/sparse.py (hunk 6 above)
Comment (tkonolige) [reply python/tvm/topi/nn/sparse.py:213]: Pylint complains these inputs are unused. Prefixing them with _ silences the warning and indicates to other people that the argument is unused.

File: src/relay/op/nn/sparse.cc (hunk 7 above)
Comment (tkonolige) [reply src/relay/op/nn/sparse.cc:106]: This op shouldn't be used directly by a user unless they are sure that the input matrix is in the correct format. I've documented this.

File: python/tvm/topi/nn/sparse.py (hunk 6 above)
Comment (ANSHUMAN87) [reply python/tvm/topi/nn/sparse.py:213]: Ok, got it. Thanks for clarifying!
We can also use macro like below 
# pylint: disable=
def func():

File: src/relay/op/nn/sparse.cc (hunk 7 above)
Comment (ANSHUMAN87) [reply src/relay/op/nn/sparse.cc:106]: I am okay with the disclaimer :)
If it is really not necessary, may be we don't expose this as global Op.
I am not too sure. Better we get some more opinion on this point.

Comment (tkonolige) [reply src/relay/op/nn/sparse.cc:106]: How can I not expose it? Do I just change the name to "internal.nn.sparse_dense_padded"?

Comment (ANSHUMAN87) [reply src/relay/op/nn/sparse.cc:106]: I think "nn.internal.xxx" is a good option.

Comment (tkonolige) [reply src/relay/op/nn/sparse.cc]: Done.

**Review by ANSHUMAN87 — APPROVED** [review]
Thanks @tkonolige !

**Review by electriclilies — APPROVED** [review]
LGTM!

**Review by jroesch — APPROVED** [review]

**Comment by merrymercy** [comment]
@tkonolige @tqchen This commits fails to be built in the master branch. see the CI: https://github.com/apache/incubator-tvm/commits/master

It introduces a flaky test that blocks two of my PRs.

**Comment by ANSHUMAN87** [comment]
> @tkonolige @tqchen This commits fails in the master branch. see the CI: https://github.com/apache/incubator-tvm/commits/master
> 
> It introduces a flaky test that blocks two of my PRs.

#6658 has resolved the issue i think.

**Comment by tkonolige** [comment]
@merrymercy I think that was the diagnostics
